XIOS implementation of `writeDiagnosticTime`

joewallwork opened the pull request:
# XIOS implementation of `writeDiagnosticTime`

Fixes #917
Fixes #929
Merges into #949

### Task List
- [x] Defined the tests that specify a complete and functioning change (*It may help to create a [design specification & test specification](../../../wiki/Specification-Template)*)
- [x] Implemented the source code change that satisfies the tests
- [x] Documented the feature by providing worked example
- [x] Updated the README or other documentation
- [x] Completed the pre-Request checklist below

---
# Change Description

This PR provides an XIOS implementation of `writeDiagnosticTime`. It completes the set of member functions needed in `ParaGridIO_Xios`, so the other copied functions are deleted and `ifdef`-ed out in the header.

Much like in #928, this required a new config section `XiosDiagnostic` (c.f. `XiosForcing`).

During my work on this, I noticed #933 and #935.

---
# Test Description

$u$-velocity field written as a diagnostic in the XIOS write test, much like how it is read in the forcing test added in #928.

Changes to XIOS read test for consistency.

---
# Documentation Impact

Diagnostics are mentioned in the XIOS section of the docs.

---
### Pre-Request Checklist

- [x] The requirements of this pull request are fully captured in an issue or design specification and are linked and summarised in the description of this PR
- [x] No new warnings are generated
- [x] The documentation has been updated (or an issue has been created to track the corresponding change)
- [x] Methods and Tests are commented such that they can be understood without having to obtain additional context
- [x] This PR/Issue is labelled as a bug/feature/enhancement/breaking change
- [x] This change conforms to the conventions described in the README

joewallwork committed:
Add XiosDiagnostic section

joewallwork committed:
Mention XiosDiagnostic in XIOS docs

joewallwork committed:
Update XIOS write test to consider diagnostics

joewallwork committed:
Similar updates for XiosRead_test

joewallwork committed:
XIOS implementation of writeDiagnosticTime

joewallwork committed:
Tidying up

joewallwork committed:
No need to setFieldFreqOffset

joewallwork committed:
Address TODOs; consistent checks before reading and writing

joewallwork committed:
Post-merge fixes

joewallwork committed:
Check filenames in ParaGridIO_Xios

joewallwork commented:
I just pushed an extra commit https://github.com/nextsimhub/nextsimdg/pull/934/commits/0e09b2cd1546378323f5f8548146b70d1fababf3 that checks the filename is correct whenever one of the reading and writing member functions of `ParaGridIO` are called. Note that this argument was previously unused (included only to ensure a consistent API with the non-XIOS version.)

TomMelt reviewed (approved):
Thanks @jwallwork23 this also looks good to go

joewallwork committed:
Introduce enum for fileType; post-rebase fixes

joewallwork committed:
Separate forcings and diagnostics files in tests

joewallwork commented:
Rebased on top of `issue938_xios-config-consistency` and changed base branch to the corresponding PR #949.

Also reworked things more neatly to handle separate files for restarts, diagnostics, and forcings.

joewallwork committed:
Merge branch 'issue938_xios-config-consistency' into issue917_xios-write-diag-time

joewallwork committed:
Merge branch 'issue938_xios-config-consistency' into issue917_xios-write-diag-time

joewallwork committed:
Merge branch 'issue938_xios-config-consistency' into issue917_xios-write-diag-time

joewallwork committed:
Post-merge fixes

joewallwork commented:
[Applied some fixes after merging in updates from #949.]

joewallwork committed:
Merge branch 'develop' into issue917_xios-write-diag-time
